qiskit/providers/ibmq/runtime/ibm_runtime_service.py

```diff
@@ -406,16 +407,46 @@ def set_program_visibility(self, program_id: str, public: bool) -> None:
                 If ``False``, make the program visible to just your account.
 
         Raises:
-            RuntimeJobNotFound: if program not found (404)
+            RuntimeProgramNotFound: if program not found (404)
             QiskitRuntimeError: if update failed (401, 403)
         """
         try:
             self._api_client.set_program_visibility(program_id, public)
         except RequestsApiError as ex:
             if ex.status_code == 404:
-                raise RuntimeJobNotFound(f"Program not found: {ex.message}") from None
+                raise RuntimeProgramNotFound(f"Program not found: {ex.message}") from None
             raise QiskitRuntimeError(f"Failed to set program visibility: {ex}") from None
 
+    def update_program(self, program_id: str, data: Union[bytes, str]) -> None:
+        """Update a runtime program's data.
+
+        Args:
+            program_id: Program ID
+            data: Name of the program file or program data to upload.
+
+        Raises:
+            FileNotFoundError: If the data filepath does not exist.
+            RuntimeProgramNotFound: If the program doesn't exist.
+            QiskitRuntimeError: If the request failed.
+            IBMQNotAuthorizedError: If not authorized to update program
+        """
+        # Convert any file data into bytes
+        if isinstance(data, str):
+            with open(data, 'r') as file:
+                data = file.read().encode()
+        # Convert the bytes data into Base64
+        data = base64.b64encode(data)
```

Comment:
Why was this logic moved to this file vs put it in `qiskit/providers/ibmq/api/rest/runtime.py`? When we put it here some statements are not covered by try/catch logic. Can this be moved back to where it was before?

Reply:
The reasons are (1) stack trace and (2) separation between REST API code from logic code. For (1), since the logic is now in the `IBMRuntimeService`, a user can probably see the point of failure a bit quicker. If you think otherwise, I'll  return it to the api function.

Reply:
Sounds good. Could we move this logic to a private function in this file itself so that it can be reused for update_program as well? We can change the update_program logic in a separate PR.

Reply:
[Updated! ](https://github.com/Qiskit/qiskit-ibmq-provider/pull/1008/files#diff-6a2aa872dcdb04786518cfee062f6f4951de5a45674504e5928c5ad3c05f459eR570-R591)Uses private function `_encode_program_data`